backend/daemon/api/documents/v1alpha/comments.go

```diff
@@ -0,0 +1,343 @@
+				docConvos = append(docConvos, obj)
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	resp := &documents.ListConversationsResponse{
+		Conversations: make([]*documents.Conversation, len(docConvos)),
+	}
+
+	for i, obj := range docConvos {
+		convo := conn.GetObjectCID(obj)
+		convpb, err := api.loadConversation(conn, convo)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load conversation: %w", err)
+		}
+
+		resp.Conversations[i] = convpb
+	}
+
+	return resp, nil
+}
+
+func (api *Server) loadConversation(conn *sqlitevcs.Conn, id cid.Cid) (*documents.Conversation, error) {
+	me := api.me.MustGet()
+
+	convo := &documents.Conversation{
+		Id: id.String(),
+	}
+
+	if err := conn.WithTx(false, func() error {
+		obj := conn.LookupPermanode(id)
+		meLocal := conn.LookupIdentity(me)
+		ver := conn.GetVersion(obj, "main", meLocal)
+		cs := conn.ResolveChangeSet(obj, ver)
+
+		sels := conn.QueryValuesByAttr(obj, cs, vcs.RootNode, mttdoc.AttrConvSelector)
+		for _, sel := range sels.Slice() {
+			selNode := sel.Value.(vcs.NodeID)
+			selBlock := conn.QueryLastValue(obj, cs, selNode, mttdoc.AttrConvSelectorBlockID)
+			if selBlock.IsZero() {
+				return fmt.Errorf("failed to find block on the selector in conversation: %s", id.String())
+			}
+			selRev := conn.QueryLastValue(obj, cs, selNode, mttdoc.AttrConvSelectorBlockRevision)
+			if selRev.IsZero() {
+				return fmt.Errorf("failed to find block revision on the selector in conversation: %s", id.String())
+			}
+
+			selpb := &documents.Selector{
+				BlockId:       selBlock.Value.(string),
+				BlockRevision: selRev.Value.(cid.Cid).String(),
+			}
+			convo.Selectors = append(convo.Selectors, selpb)
+
+			selStart := conn.QueryLastValue(obj, cs, selNode, mttdoc.AttrConvSelectorStart)
+			// If we don't have selector start range, we won't have the end either.
+			if selStart.IsZero() {
+				continue
+			}
+
+			selpb.Start = int32(selStart.Value.(int))
+			selEnd := conn.QueryLastValue(obj, cs, selNode, mttdoc.AttrConvSelectorEnd)
+			if selEnd.IsZero() {
+				return fmt.Errorf("failed to find select end while having selector start in conversation: %s", id.String())
+			}
+			selpb.End = int32(selEnd.Value.(int))
+		}
+
+		blocks := conn.QueryValuesByAttr(obj, cs, vcs.RootNode, mttdoc.AttrBlockSnapshot)
+		for _, row := range blocks.Slice() {
+			blk := decodeBlock(conn, obj, row.OpID(), row.Value.([]byte))
+			convo.Comments = append(convo.Comments, blk)
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return convo, nil
+}
+
+func validateSelectors(name string, in []*documents.Selector) error {
+	if len(in) == 0 {
+		return errutil.NotNil(name)
+	}
+
+	for _, sel := range in {
+		if sel.BlockId == "" {
+			return errutil.MissingArgument(name + ".blockID")
+		}
+
+		if sel.BlockRevision == "" {
+			return errutil.MissingArgument(name + ".blockRevision")
+		}
+
+		if sel.Start > sel.End {
+			return errutil.Greater(name+".end", sel.End, sel.Start)
+		}
```

Comment:
What about negative ofssets? in protobuf, they are defined as `int32 start = 3;`...

Reply:
Answered here: https://github.com/mintterteam/mintter/pull/1249#discussion_r1067996281